File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"

int randint(int seed,int counter){
    int a = 126354;
    int k = 1000;
    int m = (seed+a)%k;
    return (m+counter)%6 + 1;
}

int ability(void){
    time_t now = time(NULL);              
    struct tm *local = localtime(&now);   
    int seconds = local->tm_sec;  
    int cntr = 0;
    int num1 = randint(seconds,cntr);
    cntr++; 
    int num2 = randint(seconds,cntr);
    cntr++; 
    int num3 = randint(seconds,cntr);
    cntr++; 
    int num4 = randint(seconds,cntr);
    int arr[] = {num1,num2,num3,num4};
    int min = arr[0];
    int i = 1;
    while (i<4){
        if (arr[i]<min){
            min = arr[i];
        }
        i++;
    }
    return num1+num2+num3+num4-min;
}

int modifier(int score){
    float s = (float) score;
    s = s - 10;
    s = s/2;
    return (int)floor(s);
}

dnd_character_t make_dnd_character(void){
    dnd_character_t a;
    a.strength = ability();
    a.dexterity = ability();
    a.constitution = ability();
    a.intelligence = ability();
    a.wisdom = ability();
    a.charisma = ability();
    a.hitpoints = 10 + modifier(a.constitution);
    return a;
}

File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.List;
import java.util.ArrayList;
import java.util.Random;
import java.lang.Math;

class DnDCharacter {
    int strength;
    int dexterity;
    int constitution;
    int intelligence;
    int wisdom;
    int charisma;
    int modifier;
    int hitpoints;

    int ability(List<Integer> scores) {
        int s = 0;
        int smallest = scores.get(0);
        for(int i =0;i<scores.size();i++){
            if (scores.get(i)<smallest){
                smallest = scores.get(i);
            }
            s+=scores.get(i);
        }
        return s - smallest;
    }

    List<Integer> rollDice() {
        Random random = new Random();
        List<Integer> a = new ArrayList<>();
        a.add(random.nextInt(6)+1);
        a.add(random.nextInt(6)+1);
        a.add(random.nextInt(6)+1);
        a.add(random.nextInt(6)+1);
        return a;
    }

    int modifier(int input) {
        return (int)Math.floor(((float)input-10)/2);
    }

    int getStrength() {
        if (this.strength==0){
            this.strength = ability(rollDice());
        }
        return this.strength;
    }

    int getDexterity() {
        if (this.dexterity==0){
            this.dexterity = ability(rollDice());
        }
        return this.dexterity;
    }

    int getConstitution() {
        if (this.constitution==0){
            this.constitution = ability(rollDice());
        }
        return this.constitution;
    }

    int getIntelligence() {
        if (this.intelligence==0){
            this.intelligence = ability(rollDice());
        }
        return this.intelligence;
    }

    int getWisdom() {
        if (this.wisdom==0){
            this.wisdom = ability(rollDice());
        }
        return this.wisdom;
    }

    int getCharisma() {
        if (this.charisma==0){
            this.charisma = ability(rollDice());
        }
        return this.charisma;
    }

    int getHitpoints() {
        return 10 + modifier(this.constitution);
    }
}
